### ISSUE TITLE
5.7 toolchain from 5. July reports new "function call causes an infinite recursion" warnings for deprecated methods in swift-nio

### ISSUE THREAD

**Issue by dnadoba** [apple/swift#59909]
**Describe the bug**
The Swift 5.7 toolchain from 5. July for Xcode and Ubuntu produce 7 new warnings:
```
/src/Sources/NIOCore/EventLoopFuture+Deprecated.swift:19:21: warning: function call causes an infinite recursion
        return self.flatMap(callback)
...
```
Method definition looks like:
```swift
    @inlinable
    @available(*, deprecated, message: "Please don't pass file:line:, there's no point.")
    public func flatMap<NewValue>(file: StaticString = #file, line: UInt = #line, _ callback: @escaping (Value) -> EventLoopFuture<NewValue>) -> EventLoopFuture<NewValue> {
        return self.flatMap(callback)
    }
```
All methods can be found here: https://github.com/apple/swift-nio/blob/f1d62128c4587a06bee9f5455eef282694b71288/Sources/NIOCore/EventLoopFuture%2BDeprecated.swift#L16-L60

**Steps To Reproduce**
- Clone swift-nio from https://github.com/apple/swift-nio.git
- open Package.swift with Xcode and build the package (or just `NIOCore`)

Or with docker:
```
git clone https://github.com/apple/swift-nio
cd swift-nio
docker run --rm --volume "$(pwd)/:/src" --workdir "/src/" swiftlang/swift:nightly-5.7-focal swift build
```

**Expected behavior**
No warnings should be emitted

**Screenshots**
<img width="1298" alt="Screen Shot 2022-07-06 at 14 05 38" src="https://user-images.githubusercontent.com/5688874/177546153-b9f22dec-240c-4a08-ae3c-a793647df224.png">

**Environment (please fill out the following information)**
 - OS: 13.0 Beta (22A5266r)
 - Xcode Version 14.0 beta 3 (14A5270f) with Swift 5.7 Development Snapshot 2022-07-05 (a)

or Docker with swiftlang/swift:nightly-5.7-focal on arm64

**Additional context**
Didn't produce these warnings in previous 5.7 snapshots

Other issues new in the 5.7 2022-07-05 snapshot:

https://github.com/apple/swift/issues/59910
https://github.com/apple/swift/issues/59911
https://github.com/apple/swift/issues/59913

**Comment by slavapestov** [comment]
@xedin Did something change in overload resolution here? It appears we're picking the default arguments overload, which causes the infinite recursion.

**Comment by xedin** [comment]
I'm not aware of any ranking changes but I think this is could be related to the recent changes @DougGregor landed for `@preconcurrency` since all these `flatMap*` methods have an overload `@preconcurrency+@Sendable` i.e.:

```
struct Future<T> {
  func flatMap<NewValue>(_ callback: @escaping (T) -> Future<NewValue>) -> Future<NewValue> {
    fatalError()
  }

  @inlinable
  @preconcurrency
  func flatMap<NewValue>(_ callback: @escaping @Sendable (T) -> Future<NewValue>) -> Future<NewValue> {
    fatalError()
  }
}

extension Future {
  @inlinable
  @available(*, deprecated, message: "")
  func flatMap<NewValue>(file: StaticString = #file, line: UInt = #line, _ callback: @escaping (T) -> Future<NewValue>) -> Future<NewValue> {
    return self.flatMap(callback)
  }
}
```

**Comment by DougGregor** [comment]
The specific example is something like this:
```swift
struct Future<T> {
  @inlinable
  @preconcurrency
  func flatMap<NewValue>(_ callback: @escaping @Sendable (T) -> Future<NewValue>) -> Future<NewValue> {
    fatalError()
  }
}

extension Future {
  @inlinable
  @available(*, deprecated, message: "")
  func flatMap<NewValue>(file: StaticString = #file, line: UInt = #line, _ callback: @escaping (T) -> Future<NewValue>) -> Future<NewValue> {
    return self.flatMap(callback)
  }
}
```

The issue here is that, when calling the `@preconcurrency` version of `flatMap` from the deprecated one, there is an argument conversion that adds `@Sendable`, and that makes the `@preconcurrency` version worse than the deprecated one. We have ideas to address this.

**Comment by dnadoba** [comment]
Fixed in the latest 5.7 toolchain from July 17, 2022
